nearswap/src/twap.rs

```diff
@@ -0,0 +1,576 @@
+                    end = mid;
+                } else {
+                    start = mid + 1;
+                }
+            }
+            if start >= max_length {
+                start = res;
+            }
+        }
+
+        return start;
+    }
+
+    // function which will calculate mean using Mean enum
+    pub fn calculate_mean(
+        &self,
+        time: Mean,
+        max_length: usize,
+    ) -> (u128, u128) {
+        let time_diff: u64 = match time {
+            Mean::M_1MIN => to_nanoseconds(60), // 1 minute in nanoseconds
+            Mean::M_5MIN => to_nanoseconds(300), // 5 minute in nanoseconds
+            Mean::M_1H => to_nanoseconds(60 * 60),
+            Mean::M_12H => to_nanoseconds(12 * 60 * 60),
+            _ => 0
+        };
+        let last_index = self.current_idx;
+        let req_timestamp = self.observations[last_index].block_timestamp - time_diff;
+        let left_index = self.binary_search(max_length, req_timestamp);
+        
+        if left_index == 0 {
+            let total_observe = self.observations[last_index].num_of_observations;
+            let price1cumu = self.observations[last_index].price1_cumulative;
+            let price2cumu = self.observations[last_index].price2_cumulative;
+            let mean1 = price1cumu / total_observe;
+            let mean2 = price2cumu / total_observe;
+            return (mean1, mean2);
+        } else {
+            let total_observe = self.observations[last_index].num_of_observations
+                                    - self.observations[left_index - 1].num_of_observations;
+            let price1cumu = self.observations[last_index].price1_cumulative
+                                    - self.observations[left_index - 1].price1_cumulative;
+            let price2cumu = self.observations[last_index].price2_cumulative
+                                    - self.observations[left_index - 1].price2_cumulative;
+            let mean1 = price1cumu / total_observe;
+            let mean2 = price2cumu / total_observe;
+            return (mean1, mean2);
+        }
+    }
+
+    pub(crate) fn log_observation(&mut self, pool: PoolInfo) {
+        // price1, price2 calculate
+        let near = u128::try_from(pool.ynear).unwrap();
+        let reserve = u128::try_from(pool.reserve).unwrap();
+        let price1: u128 = near / reserve;
+        let price2: u128 = reserve / near;
+
+        // update current index
+        if self.observations.len == 0 {
+            self.current_idx = self.initialize(env::block_timestamp(), price1, price2);
+        } else {
+            self.current_idx = self.write(
+                env::block_timestamp(),
+                price1,
+                price2,
+                MAX_LENGTH
+            );
+        }
+
+        // update mean
+        let len: usize;
+        if self.pivoted == true {
+            len = MAX_LENGTH.into();
+        } else {
+            len = self.observations.len();
```

Comment:
I think we can always use `self.observations.len()`, right?

Reply:
yes